File: CamelToSnakeCase.go
package main

import "fmt" // Imported as requested

func CamelToSnakeCase(s string) string {
	if s == "" {
		return ""
	}
	res := ""
	prevUpper := false
	for i, r := range s {
		isLower, isUpper := r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z'
		if !isLower && !isUpper {
			return s
		}
		if isUpper && prevUpper {
			return s
		}

		if isUpper && i != 0 {
			res += "_"
		}

		res += string(r)
		prevUpper = isUpper
	}
	if s[len(s)-1] >= 'A' && s[len(s)-1] <= 'Z' {
		return s
	}
	return res
}

func main() {
	fmt.Println(CamelToSnakeCase("HelloWorld"))
	fmt.Println(CamelToSnakeCase("helloWorld"))
	fmt.Println(CamelToSnakeCase("camelCase"))
	fmt.Println(CamelToSnakeCase("CAMELtoSnackCASE"))
	fmt.Println(CamelToSnakeCase("camelToSnakeCase"))
	fmt.Println(CamelToSnakeCase("hey2"))
}

File: elementary_python/cameltosnakecase.py
def is_valid_camel_case(s: str) -> bool:
    if not s:
        return False
    if not s[0].isalpha():
        return False

    for i in range(1, len(s)):
        char = s[i] 
        if char.isdigit() or not char.isalnum():
            return False
        
        if char.isupper() and s[i-1].isupper():
            return False
    return True

def camel_to_snake_case(s: str) -> str:
    if not s:
        return ""

    if not is_valid_camel_case(s):
        return s
    
    result = []
    for i, char in enumerate(s):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char)

    return "".join(result)

if __name__ == "__main__":
    print(camel_to_snake_case("HelloWorld"))
    print(camel_to_snake_case("helloWorld"))
    print(camel_to_snake_case("camelCase"))
    print(camel_to_snake_case("CAMELtoSnackCASE"))
    print(camel_to_snake_case("camelToSnakeCase"))
    print(camel_to_snake_case("hey2"))
    print(camel_to_snake_case(""))
    print(camel_to_snake_case("Hello_World"))